### PR TITLE
Add gguf support for bloom

### PR DESCRIPTION
**Pull request by VladOS95-cyber**
# What does this PR do?
Add Bloom GGUF loading support

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Link: https://github.com/huggingface/transformers/issues/33260
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
Regarding the task @SunMarc @LysandreJik @ArthurZucker .

### CODE REVIEW

**Comment by VladOS95-cyber** [comment]
Hello @SunMarc @LysandreJik @ArthurZucker! I would like to ask for a code review

File: src/transformers/convert_slow_tokenizer.py [hunk 1]
```diff
@@ -343,9 +345,12 @@ def converted(self) -> Tokenizer:
...
         )
 
-        tokenizer.pre_tokenizer = pre_tokenizers.ByteLevel(add_prefix_space=self.original_tokenizer.add_prefix_space)
+        add_prefix_space = False
+        if hasattr(self.original_tokenizer, "add_prefix_space"):
+            add_prefix_space = self.original_tokenizer.add_prefix_space
+        tokenizer.pre_tokenizer = pre_tokenizers.ByteLevel(add_prefix_space=add_prefix_space)
         tokenizer.decoder = decoders.ByteLevel()
-        if self.original_tokenizer.add_bos_token:
+        if hasattr(self.original_tokenizer, "add_bos_token") and self.original_tokenizer.add_bos_token:
```
Comment (akx) [inline src/transformers/convert_slow_tokenizer.py:353]: Simplifies to
```suggestion
        if getattr(self.original_tokenizer, "add_bos_token", False):
```

File: src/transformers/convert_slow_tokenizer.py [hunk 2]
```diff
@@ -328,9 +328,11 @@ def converted(self) -> Tokenizer:
...
 
 class GPT2Converter(Converter):
-    def converted(self) -> Tokenizer:
-        vocab = self.original_tokenizer.encoder
-        merges = list(self.original_tokenizer.bpe_ranks.keys())
+    def converted(self, vocab: Dict[str, int] = None, merges: List[Tuple[str, str]] = None) -> Tokenizer:
+        if not vocab:
+            vocab = self.original_tokenizer.encoder
+        if not merges:
+            merges = list(self.original_tokenizer.bpe_ranks.keys())
```
Comment (akx) [inline src/transformers/convert_slow_tokenizer.py:335]: `.keys()` is superfluous.
```suggestion
            merges = list(self.original_tokenizer.bpe_ranks)
```

File: src/transformers/convert_slow_tokenizer.py [hunk 3]
```diff
@@ -343,9 +345,12 @@ def converted(self) -> Tokenizer:
             )
         )
 
-        tokenizer.pre_tokenizer = pre_tokenizers.ByteLevel(add_prefix_space=self.original_tokenizer.add_prefix_space)
+        add_prefix_space = False
+        if hasattr(self.original_tokenizer, "add_prefix_space"):
+            add_prefix_space = self.original_tokenizer.add_prefix_space
```
Comment (akx) [inline src/transformers/convert_slow_tokenizer.py:350]: Simplifies to
```suggestion
        add_prefix_space = getattr(self.original_tokenizer, "add_prefix_space", False)
```

File: src/transformers/integrations/ggml.py [hunk 4]
```diff
@@ -490,11 +509,24 @@ def converted(self) -> Tokenizer:
         return tokenizer
 
 
+class GGUFBloomConverter(GPT2Converter):
+    def __init__(self, tokenizer_dict):
+        self.original_tokenizer = GGUFTokenizerSkeleton(tokenizer_dict)
```
Comment (akx) [inline src/transformers/integrations/ggml.py:514]: This seems to be missing a superclass `__init__()` call, is that on purpose?

Comment (VladOS95-cyber) [reply src/transformers/integrations/ggml.py:514]: We don't need initializer from a base class, as we don't provide any arguments like original_tokenizer, we just call converted method

File: src/transformers/convert_slow_tokenizer.py (hunk 1 above)
Comment (VladOS95-cyber) [reply src/transformers/convert_slow_tokenizer.py:353]: done

File: src/transformers/convert_slow_tokenizer.py (hunk 2 above)
Comment (VladOS95-cyber) [reply src/transformers/convert_slow_tokenizer.py:335]: done

File: src/transformers/convert_slow_tokenizer.py (hunk 1 above)
Comment (VladOS95-cyber) [reply src/transformers/convert_slow_tokenizer.py:353]: done

**Comment by VladOS95-cyber** [comment]
Hi @SunMarc @LysandreJik @ArthurZucker! This PR is ready for review. There is one thing that looks odd to me. After dequantiozation and loading the model, It genereates a wrong sequence, not as expected when using a normal pretrained model. Instead of```tensor([[59414,    15,   473,  3370,  4026,   427,  5894,   861,   473,   912, 5636]]) ```, it generate smth like `[[59414,    15,   15,  15,  15,   15,  15,   15,   15,   15, 15]]`. I cannot find a root cause of this problem, i've already checked mapping and so on several times, it should be correct. It looks like that weights are not correct but I am not sure...

**Comment by SunMarc** [comment]
> This PR is ready for review. There is one thing that looks odd to me. After dequantization and loading the model, It genereates a wrong sequence, not as expected when using a normal pretrained model. Instead oftensor([[59414,    15,   473,  3370,  4026,   427,  5894,   861,   473,   912, 5636]]) , it generates smth like [[59414,    15,   15,  15,  15,   15,  15,   15,   15,   15, 15]]. I cannot find a root cause of this problem, i've already checked mapping and so on several times, it should be correct. It looks like that weights are not correct but I am not sure...

Since the model was quantized, THis is normal that it is not behaving the same of the normal pretrained model. Dequantization doesn't recover the precision of the original model. Could you check that it behaves similarly as the original model that was converted to gguf in fp16 precision or full precision ? This way we have a way to compare the model loaded from gguf file.

**Comment by VladOS95-cyber** [comment]
> > This PR is ready for review. There is one thing that looks odd to me. After dequantization and loading the model, It genereates a wrong sequence, not as expected when using a normal pretrained model. Instead oftensor([[59414,    15,   473,  3370,  4026,   427,  5894,   861,   473,   912, 5636]]) , it generates smth like [[59414,    15,   15,  15,  15,   15,  15,   15,   15,   15, 15]]. I cannot find a root cause of this problem, i've already checked mapping and so on several times, it should be correct. It looks like that weights are not correct but I am not sure...
> 
> Since the model was quantized, THis is normal that it is not behaving the same of the normal pretrained model. Dequantization doesn't recover the precision of the original model. Could you check that it behaves similarly as the original model that was converted to gguf in fp16 precision or full precision ? This way we have a way to compare the model loaded from gguf file.

hi @SunMarc. Still odd behaviour, it does not matter which format I take. So, i debugged model_state loading and parsing logic several times and tried to compare it with working one (like llama). Everything looks good from this perspective and it seems that weights, params, config and so on are loaded correctly into the model. I'am just worring about quantized data itself. I checked the code on https://github.com/ggerganov/llama.cpp/blob/a38b884c6c4b0c256583acfaaabdf556c62fabea/convert_hf_to_gguf.py L801-846 and it maps **bloom-style qkv_linear to gpt-style qkv_linear** by reshaping data. I am just thinking, could it be a problem and we have to reshape it back or smth? Because, honestly, I cannot find a root cause of this strange behaviour.

**Comment by VladOS95-cyber** [comment]
Hi @SunMarc! I finally found the issue and the reason of bloom model strange behaviour. This is because of https://github.com/ggerganov/llama.cpp/blob/master/convert_hf_to_gguf.py (L972-998), I honestly have no idea for what purpose it is done there but algorithm reshapes qkv data and that's why we got completely different weights in the end. I implemented reverse reshaping algorithm in modeling_gguf_pytorch_utils.py to place them back and after that, everything is ok, and model outputs expected values. Please, take a look on my changes, now, everything should be correct.

**Review by SunMarc — APPROVED** [review]
Nice job finding the issue ! Could you add a final test to check that the fp16 from transformers and the fp16 model from gguf share the same weights ? That would be a nice test to check that the conversion was done correctly ! Can you also check that the quantized model doesn't return gibberish ?

**Comment by VladOS95-cyber** [comment]
> Nice job finding the issue ! Could you add a final test to check that the fp16 from transformers and the fp16 model from gguf share the same weights ? That would be a nice test to check that the conversion was done correctly ! Can you also check that the quantized model doesn't return gibberish ?

Yes, sure, I added one more test to compare weights. But what about gibberish? Should I implement additional test for that to check particular promt or?

**Comment by SunMarc** [comment]
> Yes, sure, I added one more test to compare weights. But what about gibberish? Should I implement additional test to check particular generation or?

Just add a simple generation check for a q4 quants for example, just like what was done for other models !

**Comment by VladOS95-cyber** [comment]
> > Yes, sure, I added one more test to compare weights. But what about gibberish? Should I implement additional test to check particular generation or?
> 
> Just add a simple generation check for a q4 quants for example, just like what was done for other models !

But this kind of test is already added, it is called test_bloom_f16. Or is it not enough?

**Comment by SunMarc** [comment]
> But this kind of test is already added, it is called test_bloom_f16. Or is it not enough?

**Comment by VladOS95-cyber** [comment]
> > But this kind of test is already added, it is called test_bloom_f16. Or is it not enough?
> 
> But we need one for a quantized model since most users uses these models. With fp16, we can't be sure that the dequantize step was done correctly.

Hello! Ok, got it, I just added q8_0 test as well. Would it be sufficient?

**Review by LysandreJik — APPROVED** [review]
Awesome, thank you @VladOS95-cyber!

**Pull request merged**